### PR TITLE
add support for charliecloud container engine

### PR DESCRIPTION
**Pull request by phue**
This supersedes https://github.com/nextflow-io/nextflow/pull/1712 (I rebased against master and cherry-picked the commits by @tbugfinder in an attempt to make the DCO bot happy)

### CODE REVIEW

**Comment by pditommaso** [comment]
Was the tested with a real pipeline execution?

**Comment by phue** [comment]
Yes, I did some small-scale tests on slurm and it works.

**Comment by pditommaso** [comment]
This is awesome! we can merge then!

**Comment by tbugfinder** [comment]
thank you @phue !

**Comment by pditommaso** [comment]
Broken build ! 

```
ContainerHandlerTest > test normalize method for singularity > nextflow.container.ContainerHandlerTest.test normalize method for singularity [IMAGE: shub://busybox, NORMALIZED: shub://busybox, X: 1, EXPECTED: /path/to/busybox, #6] FAILED
    java.lang.IllegalStateException at ContainerHandlerTest.groovy:189
        Caused by: org.codehaus.groovy.runtime.InvokerInvocationException at SingularityCache.groovy:281
            Caused by: java.io.IOException at SingularityCache.groovy:143
ContainerHandlerTest > test normalize method for singularity > nextflow.container.ContainerHandlerTest.test normalize method for singularity [IMAGE: docker://library/busybox, NORMALIZED: docker://library/busybox, X: 1, EXPECTED: /path/to/busybox, #7] FAILED
    java.lang.IllegalStateException at ContainerHandlerTest.groovy:189
        Caused by: org.codehaus.groovy.runtime.InvokerInvocationException at SingularityCache.groovy:281
            Caused by: java.io.IOException at SingularityCache.groovy:143
ContainerHandlerTest > test normalize method for singularity > nextflow.container.ContainerHandlerTest.test normalize method for singularity [IMAGE: foo, NORMALIZED: docker://foo, X: 1, EXPECTED: /path/to/foo, #8] FAILED
    java.lang.IllegalStateException at ContainerHandlerTest.groovy:189
        Caused by: org.codehaus.groovy.runtime.InvokerInvocationException at SingularityCache.groovy:281
            Caused by: java.io.IOException at SingularityCache.groovy:143
ContainerHandlerTest > test normalize method for singularity > nextflow.container.ContainerHandlerTest.test normalize method for singularity [IMAGE: library://pditommaso/foo/bar.sif:latest, NORMALIZED: library://pditommaso/foo/bar.sif:latest, X: 1, EXPECTED: /path/to/foo-bar-latest.img, #9] FAILED
    java.lang.IllegalStateException at ContainerHandlerTest.groovy:189
        Caused by: org.codehaus.groovy.runtime.InvokerInvocationException at SingularityCache.groovy:281
            Caused by: java.io.IOException at SingularityCache.groovy:143
ContainerHandlerTest > test normalize method for singularity FAILED
ContainerHandlerTest > should not invoke caching when engine is disabled FAILED
    java.lang.IllegalStateException at ContainerHandlerTest.groovy:225
        Caused by: org.codehaus.groovy.runtime.InvokerInvocationException at SingularityCache.groovy:281
            Caused by: java.io.IOException at SingularityCache.groovy:143
ContainerHandlerTest > should invoke singularity cache > nextflow.container.ContainerHandlerTest.should invoke singularity cache [IMG: foo, NORM: docker://foo, TIMES: 1, EXPECTED: /local/img/foo, #0] FAILED
    java.lang.IllegalStateException at ContainerHandlerTest.groovy:237
        Caused by: org.codehaus.groovy.runtime.InvokerInvocationException at SingularityCache.groovy:281
            Caused by: java.io.IOException at SingularityCache.groovy:143
ContainerHandlerTest > should invoke singularity cache > nextflow.container.ContainerHandlerTest.should invoke singularity cache [IMG: library://foo:latest, NORM: library://foo:latest, TIMES: 1, EXPECTED: /local/img/foo.img, #1] FAILED
    java.lang.IllegalStateException at ContainerHandlerTest.groovy:237
        Caused by: org.codehaus.groovy.runtime.InvokerInvocationException at SingularityCache.groovy:281
            Caused by: java.io.IOException at SingularityCache.groovy:143
ContainerHandlerTest > should invoke singularity cache > nextflow.container.ContainerHandlerTest.should invoke singularity cache [IMG: http://bar:latest, NORM: http://bar:latest, TIMES: 1, EXPECTED: /local/http/foo.img, #2] FAILED
    java.lang.IllegalStateException at ContainerHandlerTest.groovy:237
        Caused by: org.codehaus.groovy.runtime.InvokerInvocationException at SingularityCache.groovy:281
            Caused by: java.io.IOException at SingularityCache.groovy:143
ContainerHandlerTest > should invoke singularity cache > nextflow.container.ContainerHandlerTest.should invoke singularity cache [IMG: https://bar:latest, NORM: https://bar:latest, TIMES: 1, EXPECTED: /local/https/foo.img, #3] FAILED
    java.lang.IllegalStateException at ContainerHandlerTest.groovy:237
        Caused by: org.codehaus.groovy.runtime.InvokerInvocationException at SingularityCache.groovy:281
            Caused by: java.io.IOException at SingularityCache.groovy:143
```

**Comment by phue** [comment]
> Broken build !

Ouch! Sorry about that. It's fixed now but I should probably add a similar test for charliecloud

**Comment by pditommaso** [comment]
This time failed a test involving a container but it seems related to a network problem 

```
   Command error:
     Unable to find image 'quay.io/nextflow/tests:latest' locally
     docker: Error response from daemon: received unexpected HTTP status: 502 Bad Gateway.
```

I'll try to execute it again

File: docs/charliecloud.rst [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+Nextflow integration with `Charliecloud <https://hpc.github.io/charliecloud>`_ technology allows you to write self-contained
+and truly reproducible computational pipelines.
+
+By using this feature any process in a Nextflow script can be transparently executed into a Podman container. This may
+be extremely useful to package the binary dependencies of a script into a standard and portable format that can be 
+executed on any platform supporting the Podman engine.
+
+.. note::
+    This feature requires Nextflow version ``20.01.0`` or later.
```
Comment (pditommaso) [inline docs/charliecloud.rst:15]: Can you please change this to version `20.12.0-edge`  ?

File: modules/nextflow/src/main/groovy/nextflow/executor/BashWrapperBuilder.groovy [hunk 2]
```diff
@@ -427,6 +428,8 @@ class BashWrapperBuilder {
         /*
          * create a builder instance given the container engine
          */
+        if( engine == 'charliecloud' )
+            return new CharliecloudBuilder(containerImage)
         if( engine == 'docker' )
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/executor/BashWrapperBuilder.groovy:433]: One minor thing, let's move this as the last. Sorry, last arrived, last served 😄

File: docs/charliecloud.rst (hunk 1 above)
Comment (phue) [reply docs/charliecloud.rst]: Done, I also updated the docs.

File: modules/nextflow/src/main/groovy/nextflow/executor/BashWrapperBuilder.groovy (hunk 2 above)
Comment (phue) [reply modules/nextflow/src/main/groovy/nextflow/executor/BashWrapperBuilder.groovy:433]: The section looked almost alphabetical :sweat_smile: 
I changed it

**Comment by pditommaso** [comment]
All green! 🎉

**Comment by pditommaso** [comment]
Look @ewels! Only 21 hours 😆

**Comment by phue** [comment]
Awesome! Thanks @pditommaso 
:tada:

**Comment by ewels** [comment]
> Look @ewels! Only 21 hours 😆

haha, and this one was much bigger than just a few lines of a docs file! 😆

**Pull request merged**